### PR TITLE
blueprints/api

### LINKED ISSUE #84: Design and implement data model for blueprints

### PR DESCRIPTION
**Pull request by timperrett**
resolves #84 and #87

### CODE REVIEW

**Comment by timperrett** [comment]
Implements #84 and #87

File: core/src/main/scala/Blueprint.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  def parseNamedRevision(serialized: String): Either[Throwable,(String, Revision)] =
+    serialized.split('@') match {
+      case Array(name, "HEAD") => Right((name, Blueprint.Revision.HEAD))
+      case Array(name, rstr)   => Revision.fromString(rstr).map(x => (name,x))
+      case Array(name)         => Right((name, Blueprint.Revision.HEAD))
+    }
+
+  sealed trait Revision
+  object Revision {
+    final object HEAD extends Revision
```
Comment (adelbertc) [inline core/src/main/scala/Blueprint.scala]: Will need `case object` so we can pattern match on this

File: core/src/main/scala/Blueprint.scala [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    def fromString(s: String): Either[Throwable,Revision] =
+      if (s.toUpperCase == HEAD.toString) Right(HEAD)
+      else Either.catchNonFatal(s.toLong).map(Discrete)
+  }
+
+  sealed trait State
+  object State {
+    final object Pending extends State {
```
Comment (adelbertc) [inline core/src/main/scala/Blueprint.scala]: case objects

File: core/src/main/resources/db/migration/V2_1__add_blueprints.sql [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ALTER TABLE IF EXISTS PUBLIC."blueprints"
+  ADD CONSTRAINT PUBLIC."blueprints_pk" PRIMARY KEY("id");
+
+ALTER TABLE IF EXISTS "deployments"
+  ADD "blueprint_id" BIGINT;
+
+ALTER TABLE PUBLIC."deployments"
+  ADD CONSTRAINT PUBLIC."deployments_have_blueprints"
+  FOREIGN KEY("blueprint_id")
+  REFERENCES PUBLIC."blueprints"("id") NOCHECK;
```
Comment (adelbertc) [inline core/src/main/resources/db/migration/V2_1__add_blueprints.sql]: We also want the fully interpolated blueprint too right, per our discussion earlier today? Needed for deleting and debugging deployments

**Pull request merged**